addons/web_tour/static/src/js/tip.js

```diff
@@ -110,19 +122,86 @@ var Tip = Widget.extend({
-            this.$anchor = $anchor;
-            this.$ideal_location = this._get_ideal_location();
-            if (this.$el.parent()[0] !== document.body) {
-                this.$el.appendTo(this.$ideal_location);
+        this._unbind_anchor_events();
+        this._setupAnchor($anchor);
+        this._updatePosition();
+        this._reposition();
+        this._bind_anchor_events();
+    },
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    _setupAnchor: function ($anchor) {
+        this.$anchor = $anchor;
+        this.$actualAnchor = this.$anchor;
+        this.$ideal_location = this._get_ideal_location();
+        this.$furtherIdealLocation = this._get_ideal_location(this.$ideal_location);
+    },
+    /**
+     * This function figure out which direction the tip should take
+     * and if he's at the bottom or the top of the targeted element
+     * or if it's an indicator to scroll.
+     */
+    _updatePosition: function () {
+        this.moveTip = false;
+        let halfHeight = 0;
+        if (this.initialPosition === 'right' || this.initialPosition === 'left') {
+            halfHeight = this.$anchor.innerHeight() / 2;
+        }
+        let position = this.info.position;
+        /* Check if the viewport state change to know if we need to move the anchor of the tip.
+        * up : the target element is above the current viewport
+        * down : the target element is below the current viewport
+        * in : the target element is in the current viewport
+        */
+        let viewPortState = "in";
+        const paddingTop = parseInt(this.$ideal_location.css('padding-top'));
+        if (this.$anchor.offset().top + this.$anchor.innerHeight() - paddingTop - halfHeight < window.pageYOffset) {
+            viewPortState = 'up';
+            position = "bottom";
+        } else if (this.$anchor.offset().top + halfHeight > window.pageYOffset + window.innerHeight) {
+            viewPortState = 'down';
+            position = "top";
+        }
+        if (this.viewPortState !== viewPortState) { // If the state change, allow to move the tip at the right place
+            this.viewPortState = viewPortState;
+            this.moveTip = true;
+        }
+
+        // Adjust the placement of the tip regarding its anchor depending
+        // if we came from the bottom or the top.
+        if (this.$anchor.offset().top < window.pageYOffset + paddingTop + this.$el.innerHeight()) {
+            position = halfHeight && viewPortState === "in" ? this.initialPosition : "bottom";
+        } else if (this.$anchor.offset().top > window.pageYOffset + window.innerHeight - (this.$anchor.innerHeight() + this.$el.innerHeight())) {
+            position = halfHeight && viewPortState === "in" ? this.initialPosition : "top";
+        }
+        this.moveTip = this.info.position !== position || this.moveTip;
```

Comment:
The calculations aren't really the same, the first **if else if** about the viewport. Does my anchor is in the viewport or not ? So does my tip is link to it or just tell me to scroll ?

The second one is only when the viewport is "in" (Maybe I can insert this **if else if** in the first one **else**), and he checks if the tip is visible (In the case the anchor isn't entirely in the viewport, the tip should be linked to the side which is in the viewport). 

For the this.moveTip, I use it in the `_reposition()`. If I scroll when the tip is open and that makes the tip going out of the viewport, the tip will not be correctly moved because the first condition in _reposition() prevent its behavior. But it should be move, because now, the tip is no longer linked to the anchor but ask the user to scroll.

Reply:
- this.$anchor.offset().top - paddingTop is written twice, no matter their goal, those are the same calculation.
- And like I said, I removed moveTip in _reposition and saw no difference, even when the tip is opened, so does it really have an effect?

Reply:
Like I said, the only case when the moveTip is useful in _reposition is when you scroll when a tip is open. So the anchor go out of viewport and we should reposition it (but the first line in _reposition prevent it because the tip is open).
I agree that's a weird case, I just tried to "limit" test the tip thing, but it can happen to a client in multiple situations.

Reply:
> I removed moveTip in _reposition and saw no difference, **even when the tip is opened**

So if I retry the case you explain tomorrow without the moveTip in the if, it will not work. Ok, I'll try again.

Reply:
I can show you my screen, but I clearly have a bug on my side.